ir/visitor.cpp

```diff
@@ -20,76 +20,68 @@ limitations under the License.
+    /* Begin tracking `n` during a visiting pass.  Use `finish` to mark `n` as
+     * visited once the pass completes.
+    */
+    void start(const IR::Node *n) {
+        // Initialization
+        visited_t::iterator visited_it;
+        bool inserted;
+        std::tie(visited_it, inserted) =
+            visited.emplace(n, visit_info_t({true,n}));
+
+        // Sanity check for IR loops
+        bool already_present = !inserted;
+        visit_info_t *visit_info = &(visited_it->second);
+        if (already_present && visit_info->visit_in_progress)
+            BUG("IR loop detected "); 
     }
-    const IR::Node *result(IR::Node *n) const { return visited.at(n).result; }
 
-    change_t track(const IR::Node *n) { return change_t(&visited, n); }
-    void start(change_t &change) { change.visit_info->visit_finished = false; }
-    bool finish(change_t &change, const IR::Node *orig, const IR::Node *final) {
+    /* Mark `n` as finished.  `done(n)` will return true, and `result(n)` will
+     * return the resulting node, if any. */
+    bool finish(const IR::Node *orig, const IR::Node *final) {
         auto it = visited.find(orig);
-        if (!change.valid || it == visited.end() || change.original_node != it->first)
+        if (it == visited.end())
             BUG("visitor state tracker corrupted");
-        change.visit_info->visit_finished = true;
+
+        visit_info_t *visit_info = &(it->second);
+        visit_info->visit_in_progress = false;
         if (!final || *final != *orig) {
-            change.visit_info->result = final;
-            visited.emplace(final, visit_info_t({true,final}));
+            visit_info->result = final;
+            visited.emplace(final, visit_info_t({false,final}));
             return true;
         } else {
             // FIXME -- not safe if the visitor resurrects the node (which it shouldn't)
             // if (final && final->id == IR::Node::currentId - 1)
             //     --IR::Node::currentId;
             return false; } }
-    const IR::Node *result(const IR::Node *n) const {
-        auto it = visited.find(n);
-        if (it == visited.end())
-            return n;
-        if (!it->second.visit_finished)
-            BUG("IR loop detected");
-        return it->second.result; }
+
+    /* Forget nodes that have already been visited, allowing them to be visited
+     * again. */
     void revisit_visited() {
         for (auto it = visited.begin(); it != visited.end();) {
-            if (it->second.visit_finished)
+            if (!it->second.visit_in_progress)
                 it = visited.erase(it);
             else
                 ++it; } }
+
+    /* False if n has never been visited or is currently being visited.
+     * True if n has been visited and the visitor has finished. */
+    bool done(const IR::Node *n) const {
+        return visited.count(n) && !visited.at(n).visit_in_progress;
+    }
+
+    /* Fails with out_of_range exception if n has not been tracked. */
```

Comment:
Here too. It'd be good to describe the behavior of this method and not just how it may fail.